sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -2586,6 +2586,22 @@ object SQLConf {
       .checkValue(_ > 0, "The timeout value must be positive")
       .createWithDefault(10L)
 
+  val LEGACY_NUMERIC_CONVERT_TO_TIMESTAMP_ENABLE =
+    buildConf("spark.sql.legacy.numericConvertToTimestampEnable")
+      .doc("when true,use legacy numberic can convert to timestamp")
+      .version("3.0.0")
+      .booleanConf
+      .createWithDefault(false)
+
+  val LEGACY_NUMERIC_CONVERT_TO_TIMESTAMP_IN_SECONDS =
+    buildConf("spark.sql.legacy.numericConvertToTimestampInSeconds")
+      .internal()
+      .doc("The legacy only works when LEGACY_NUMERIC_CONVERT_TO_TIMESTAMP_ENABLE is true." +
+        "when true,the value will be  interpreted as seconds,which follow spark style," +
+        "when false,value is interpreted as milliseconds,which follow hive style")
```

Comment:
Sorry but I can't still follow why Spark should follow Hive style even by default. Most likely the legacy users are already depending on this behaviour, and few users might had to do the workaround by themselves. I don't think even `cast(ts as long)` is a standard and an widely accepted behaviour. -1 from me.

Reply:
Hi HyukjinKwon
thanks for reviewing,we discussed the pain point when we move to spark in #28568,  i mean we can adopt both the compatibility  flag and adding functions, for using the function, user need to modify tasks one by one with the casting compatibility  flag turning off,unfortunally ,we have  almost hundred thousand tasks migrating from hive to spark, so with a flag ,we will first fail the tasks if it had CAST_NUMERIC_TO_TIMESTAMP,if user really want to use, we will suggest the NEW adding three functions for them,maybe it's a good way to avoid the case when the task has been succeed,while the casting result is wrong,which is more serious,maybe other brothers meet the same headache problem,so i hope we will embracing spark better with this patch,

Reply:
This kind of compatibility isn't fully guaranteed in Spark, see also [Compatibility with Apache Hive](https://spark.apache.org/docs/latest/sql-migration-guide-hive-compatibility.html). Simply following Hive doesn't justify this PR.

There are already a bunch of mismatched behaviours and I don't like to target more compatibility, in particular, by fixing the basic functionalities such as cast and adding such switches to maintain. Why is it difficult to just add `ts / 1000`? The workaround seems very easy.

If we target to get rid of `cast(ts as long)` away, adding separate functions is a-okay because it doesn't affect existing users, and also looks other DBMSs have their own ways by having such functions in general. Looks we will have workarounds once these functions from https://github.com/apache/spark/pull/28534 are merged, and seems you can leverage these functions alternatively as well.

I would say a-no to fix a basic functionality to have another variant and non-standard behaviour, which could potentially trigger to have another set of non-standard behaviours in many other places in Spark.

Reply:
Hi HyukjinKwon
i agree with you that the function is a cool desgin ,But to be honest , if i change one sql,it will be easy,but we have almost hundred thousand tasks,To change all tasks with obvious  conversion,implicit conversion，or expression convesion  will be a huge task,some individual have over 1 thousand tasks, it will be a really touch thing.i will jump for joy if we have  better ideas . @HyukjinKwon

Reply:
There's no such workload that can be migrated without touching anything in practice from A to B system where A doesn't guarantee full compatibility with B. I don't have a good idea for your workload.

I don't think this is only the case where it needs some fixes when you migrate from Hive to Spark. Spark doesn't target the full compatibility by design. We could think about some non-invasive fixes practically but this fix seems no.

Reply:
@cloud-fan how do you think?

Reply:
FWIW, I think there's a viable solution for you that doesn't involve changes to Spark.

1. You add a UDF to Hive for TIMESTAMP_MILLIS().
2. You search all your users' workloads for "CAST(... AS TIMESTAMP)" (case insensitive, allowing multiple spaces and newlines, allowing multiline matches) and migrate them to TIMESTAMP_MILLIS(). For most common cases this can likely be done with an "sed" script, assuming the workloads are in files.

After that, you have a workload that will run in both systems.

I think forbidding the cast in Spark is a good thing and it could be used as a safety net (i.e., if you missed a cast in a query somewhere then it'll fail in Spark), but that's not required here.

Reply:
+1 to simply forbid cast long to timestamp in Spark. Hive compatibility is not strong enough to justify the change, as other people may keep adding new behaviors for compatibility with other systems, and this can be end-less.

Instead, I think it's better to forbid this non-standard cast. You can find all the places that need to change, with explicit error from Spark. And you can add Hive UDF as @bart-samwel suggested, if you need to fallback to Hive.

Reply:
+1 from me too

Reply:
Great!  it seems i cannot reopen the pr

Reply:
reopened

Reply:
Thanks @cloud-fan ,i have reverted commit, could you help review ?